File: contract/extract.py
import os
import re

contracts = ['arena', 'factory', 'payout', 'rwa-adapter', 'oracle', 'staking']
output = "# Smart Contracts API Documentation\n\n"

for c in contracts:
    if not os.path.isdir(f"contract/{c}"):
        continue
    output += f"## {c.capitalize()}\n\n"
    
    # Extract storage keys
    try:
        with open(f"contract/{c}/src/storage.rs", 'r', encoding='utf-8') as f:
            content = f.read()
            keys = re.findall(r'pub enum DataKey \{(.*?)\}', content, re.DOTALL)
            if keys:
                output += "### Storage Keys\n"
                for k in [x.strip() for x in keys[0].split(',') if x.strip()]:
                    output += f"- `{k}`\n"
                output += "\n"
    except Exception as e:
        pass
    
    # Extract errors
    try:
        with open(f"contract/{c}/src/types.rs", 'r', encoding='utf-8') as f:
            content = f.read()
            errors = re.findall(r'pub enum .*?Error \{(.*?)\}', content, re.DOTALL)
            if errors:
                output += "### Errors\n"
                for e in [x.strip() for x in errors[0].split(',') if x.strip()]:
                    output += f"- `{e}`\n"
                output += "\n"
    except Exception:
        try:
            with open(f"contract/{c}/src/lib.rs", 'r', encoding='utf-8') as f:
                content = f.read()
                errors = re.findall(r'pub enum .*?Error \{(.*?)\}', content, re.DOTALL)
                if errors:
                    output += "### Errors\n"
                    for e in [x.strip() for x in errors[0].split(',') if x.strip()]:
                        output += f"- `{e}`\n"
                    output += "\n"
        except Exception:
            pass

    # Extract pub functions
    try:
        with open(f"contract/{c}/src/lib.rs", 'r', encoding='utf-8') as f:
            content = f.read()
            impl_blocks = re.findall(r'#\[contractimpl\]\s*impl.*?\{(.*?)\n\}', content, re.DOTALL)
            output += "### Functions\n"
            for b in impl_blocks:
                fns = re.findall(r'pub fn (\w+)\((.*?)\)(?:\s*->\s*(.*?))?\s*\{', b)
                for name, args, ret in fns:
                    args_clean = " ".join(args.split())
                    ret_str = f" -> {ret.strip()}" if ret else ""
                    output += f"- `pub fn {name}({args_clean}){ret_str}`\n"
            output += "\n"
    except Exception:
        pass

with open("contract/CONTRACTS.md", "w", encoding='utf-8') as f:
    f.write(output)

File: contract/extract.js
const fs = require('fs');
const path = require('path');

const contracts = ['arena', 'factory', 'payout', 'rwa-adapter', 'oracle', 'staking'];
let output = "# Smart Contracts API Documentation\n\n";

for (const c of contracts) {
    const dir = path.join('contract', c, 'src');
    if (!fs.existsSync(dir)) continue;

    output += `## ${c.charAt(0).toUpperCase() + c.slice(1)}\n\n`;

    // Extract storage keys
    try {
        const storagePath = path.join(dir, 'storage.rs');
        if (fs.existsSync(storagePath)) {
            const content = fs.readFileSync(storagePath, 'utf8');
            const match = content.match(/pub enum DataKey\s*\{([\s\S]*?)\}/);
            if (match) {
                output += "### Storage Keys\n";
                const keys = match[1].split(',').map(k => k.trim()).filter(k => k.length > 0);
                for (const k of keys) {
                    output += `- \`${k}\`\n`;
                }
                output += "\n";
            }
        }
    } catch (e) {}

    // Extract errors
    try {
        let typesPath = path.join(dir, 'types.rs');
        if (!fs.existsSync(typesPath)) {
            typesPath = path.join(dir, 'lib.rs');
        }
        if (fs.existsSync(typesPath)) {
            const content = fs.readFileSync(typesPath, 'utf8');
            const match = content.match(/pub enum .*?Error\s*\{([\s\S]*?)\}/);
            if (match) {
                output += "### Errors\n";
                const errs = match[1].split(',').map(e => e.trim().split(/\s*=/)[0]).filter(e => e.length > 0 && !e.startsWith('//'));
                for (const e of errs) {
                    output += `- \`${e}\`\n`;
                }
                output += "\n";
            }
        }
    } catch (e) {}

    // Extract pub functions
    try {
        const libPath = path.join(dir, 'lib.rs');
        if (fs.existsSync(libPath)) {
            const content = fs.readFileSync(libPath, 'utf8');
            const implBlocks = [...content.matchAll(/#\[contractimpl\][\s\S]*?impl[\s\S]*?\{([\s\S]*?^\})/gm)];
            if (implBlocks.length > 0) {
                output += "### Functions\n";
                for (const block of implBlocks) {
                    const blockContent = block[1];
                    const fns = [...blockContent.matchAll(/pub fn\s+(\w+)\s*\((.*?)\)(?:\s*->\s*(.*?))?\s*\{/g)];
                    for (const fn of fns) {
                        const name = fn[1];
                        const args = fn[2].replace(/\s+/g, ' ').trim();
                        const ret = fn[3] ? ` -> ${fn[3].trim()}` : '';
                        output += `- \`pub fn ${name}(${args})${ret}\`\n`;
                    }
                }
                output += "\n";
            }
        }
    } catch (e) {}
}

fs.writeFileSync(path.join('contract', 'CONTRACTS.md'), output);
console.log('CONTRACTS.md generated successfully.');
